### ISSUE TITLE
Action dispatched twice if space in params and + instead of %20 in encoded url

### ISSUE THREAD

**Issue by lawnzapper** [faceyspacey/redux-first-router#183]
Hey faceyspacey, first of all thank you for your efforts with reactlandia and these amazing tools you are working on.

I'm building my urls with query params as path separated by '+'
my url should be: **example.com/keywords/my+keywords**

Using NavLink I'm encountering some issues:

**Example 1**
`
<NavLink activeClassName={styles.active} 
to='/keywords/my keywords'>
my keywords
</NavLink>
`
The action here is dispatched twice by redux-first-router, the first time to
`location.pathname: '/keywords/my keywords'`
the second time to the encoded
`location.pathname: '/keywords/my%20keywords'`
Url bar reflects the encoded pathname
This behavior also causes the resource to be fetched twice

**Example 2** ( this is what I'm trying to accomplish)
`
<NavLink activeClassName={styles.active} 
to='/keywords/my+keywords'>
my keywords
</NavLink>
`

In this case when the action is dispatched once and navigates to the encoded
`location.pathname: '/keywords/my%2Bkeywords'`
and obviously my resource will be fetched as 'my+keywords'
but I need to fetch it as 'my keywords'
to solve this I can `replace(/\+/, ' ')` in `toPath` property on my routesMap definition,
but now Url bar reflects again the encode pathname with %20 and I want to use '+' as separator in Url bar with eventually an inverted replace at `fromPath`

I see  you are using **pathToRegexp** version 1.7.0 ( in rudy-match-path and redux-first-router-link ) which doesn't have any custom options to use '+' instead of '%20%', but in latest version 2.1.0 it has an option parameter that accepts a custom function instead of encodeURIComponent ( undocumented )

Well, to solve all my issues I could hardcode my anchor tag with 'my+keywords', but then I would lose the benefit of the action based routing that redux-first-router offers.

Long story short: How can I use '+' instead of '%20' in encoded url?

**Comment by faceyspacey** [comment]
Can you link me to that undocumented custom function.

I wanna fix this bug in the upcoming Rudy launch. I find it weird that 2 actions are dispatched--i guess the browser redirects or something, causing the history change to trigger RFR again??

As soon as I get a chance I'll reproduce this for myself and take a look.

**Comment by lawnzapper** [comment]
Concerning **pathToRegexp**, 
[here you can see the current documented Options](https://github.com/pillarjs/path-to-regexp/blame/master/Readme.md#L28-L37)
and [here the custom function undocumented in Index.js on line 139](https://github.com/pillarjs/path-to-regexp/blob/master/index.js#L119-L140)

I made a [fork of redux-first-router-demo sandbox](https://po38xjo1x.codesandbox.io) to show you some examples

Clicking on the first NavLink which has 'to' property set with type and payload 
`{category: 'react redux'}` 
redux-first-router dispatch two actions, and yes it could be related to the browser, but as you can see in [browser preview in condsandbox](https://codesandbox.io/s/po38xjo1x) the url is not encoded, but again as you can see in console, thunk from 'LIST' routesMap gets logged twice. Then if you [open preview as separate tab](https://po38xjo1x.codesandbox.io/) the url gets encoded and action dispathed twice as usual.

Note that I had to made a  `replace(/\s|%20/g, '-')` both on category reducer and isLoading selector just to show videos on page. This made me think that maybe there's some confusion around category, slug or any other custom reduced state, because sometime you get category from the main state of the app ( ex. in List Component ) and sometime from the location state ( ex. in isLoading Selector ). It could end to be neither  immutable nor a single source of truth, don't you think?

Now I got confused, anyway in the sandbox you can find that the other NavLinks work as expected ( Action dispatched once ), and the last two, going back to the 'encoding issue', with '+' get encoded in [preview as separate tab](https://po38xjo1x.codesandbox.io/) and not encoded in [browser preview in condsandbox](https://codesandbox.io/s/po38xjo1x)

**Comment by faceyspacey** [comment]
one quick thing: if you use the `capitalizedWords: true` route option, it will make the payload with space work correctly without dispatching twice:

https://codesandbox.io/s/4rnwlkko90


https://github.com/faceyspacey/redux-first-router/blob/master/src/pure-utils/actionToPath.js#L42

https://github.com/faceyspacey/redux-first-router/blob/master/src/pure-utils/pathToAction.js#L52

Does that solve your problem?

**Comment by glennr** [comment]
Also running into this issue!

**Comment by faceyspacey** [comment]
what if you add: `toPath: str => encodeURIComponent(str)` and encode it yourself?

These functions are both called on dispatches, regardless of the direction (i.e. regardless if it's a dispatch or listening to URL changes). So perhaps we just gotta manually align them properly ourselves.

**Comment by lawnzapper** [comment]
Hi faceyspacey, while waiting for Respond I made some small changes to redux-first-router actionToPath and rudy-match-path. I made two pull requests, if you could have a look would be really nice.
